waspc/src/Wasp/Generator/PackageJsonGenerator.hs

```diff
@@ -1,17 +1,71 @@
 module Wasp.Generator.PackageJsonGenerator
   ( resolveNpmDeps,
+    resolveDependencies,
+    DependencyConflictError (DependencyConflictError),
     npmDepsToPackageJsonEntry,
     npmDevDepsToPackageJsonEntry,
   )
 where
 
-import Data.Bifunctor (second)
-import Data.List (find, intercalate)
-import Data.Maybe (fromJust, isJust)
+import Data.List (intercalate)
+import qualified Data.Map as Map
+import qualified Data.Maybe as Maybe
 import qualified Wasp.AppSpec.App.Dependency as D
 
 type NpmDependenciesConflictError = String
 
+data DependencyConflictError = DependencyConflictError D.Dependency D.Dependency
+  deriving (Show, Eq)
+
+-- | Takes wasp npm dependencies and user npm dependencies and figures out how
+--   to combine them together, returning (Right) a list of npm dependencies to
+--   be used on behalf of wasp and then also a list of npm dependencies to be
+--   used on behalf of user. These lists might be the same as the initial ones,
+--   but might also be different.
+
+--   On error (Left), returns a list of conflicting wasp dep, user dep pairs
+
+--   This is used in resolveNpmDeps to separate the error message generation
+--   from its representation
+resolveDependencies ::
+  [D.Dependency] ->
+  [D.Dependency] ->
+  Either
+    [DependencyConflictError]
+    ([D.Dependency], [D.Dependency])
+resolveDependencies waspDeps userDeps =
+  if null conflicting
+    then Right (waspDeps, createDeps userDepsNotInWaspDeps)
+    else Left conflicting
+  where
+    waspDepsMap = createMap waspDeps
+    userDepsMap = createMap userDeps
+    overlappingDeps = waspDepsMap `Map.intersection` userDepsMap
+    userDepsNotInWaspDeps = userDepsMap `Map.difference` waspDepsMap
+    createMap :: [D.Dependency] -> Map.Map String D.Dependency
+    createMap deps = Map.fromList $ fmap toTuple deps
+      where
+        toTuple :: D.Dependency -> (String, D.Dependency)
+        toTuple dep = (D.name dep, dep)
+    createDeps :: Map.Map String D.Dependency -> [D.Dependency]
+    createDeps depsMap = Map.elems depsMap
```

Comment:
I would consider not having this function and using Map.elems directly, or name it smth like `depsFromDepsByName`.

Reply:
Good call, I've inlined it. I really needed the types in the past but the code underwent some refactorings already and I hadn't noticed I could simplify this.